R/paginate.R

```diff
@@ -0,0 +1,331 @@
+
+  if (is.null(total) || is.null(page_size)) {
+    return(max_pages)
+  }
+
+  body_parsed <- resp_body_json(resp)
+  total <- purrr::pluck(body_parsed, total)
+  if (is.null(total)) {
+    return(max_pages)
+  }
+  n_pages <- ceiling(total / page_size)
+
+  min(n_pages, max_pages)
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_next_url <- function(req,
+                                  next_url,
+                                  ...,
+                                  page_size = NULL,
+                                  total = NULL) {
+  check_character(next_url)
+
+  next_request <- function(req, resp) {
+    body_parsed <- resp_body_json(resp)
+    next_url <- purrr::pluck(body_parsed, next_url)
+
+    if (is.null(next_url)) {
+      return(NULL)
+    }
+
+    req_url(req, next_url)
+  }
+
+  req_paginate(
+    req,
+    next_request,
+    page_size = page_size,
+    total = total
+  )
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_offset <- function(req,
+                                offset,
+                                page_size,
+                                total = NULL) {
+  check_param(offset, allow_null_value = TRUE)
+  check_param(page_size)
+
+  cur_offset <- 0L
+  env <- current_env()
+  next_request <- function(resp, req) {
+    cur_offset <- get("cur_offset", envir = env)
+    new_offset <- cur_offset + page_size$value
+    assign("cur_offset", new_offset, envir = env)
+
+    req_set_param(req, offset, new_offset)
+  }
+
+  req_paginate(
+    req,
+    next_request,
+    page_size = page_size,
+    total = total
+  )
+}
+
+#' @rdname req_paginate
+#' @export
+req_paginate_next_token <- function(req,
+                                    token_field,
+                                    next_token_field,
+                                    page_size = NULL,
+                                    total = NULL) {
+  check_param(token_field, allow_null_value = TRUE)
+  check_character(next_token_field)
+
+  next_request <- function(req, resp) {
+    body_parsed <- resp_body_json(resp)
+    next_token <- purrr::pluck(body_parsed, next_token_field)
+
+    if (is.null(next_token)) {
+      return(NULL)
+    }
+
+    req_set_param(req, token_field, next_token)
+  }
+
+  req_paginate(
+    req,
+    next_request,
+    page_size = page_size,
+    total = total
+  )
+}
+
+#' Specify a request parameter
+#'
+#' @param name Name of the query resp. header parameter.
+#' @param path Path of the parameter in the body.
+#' @param value The value of the parameter.
+#' @param error_call
+#'
+#' @return A parameter object.
+#' @export
+#'
+#' @examples
+#' in_query("start", 20)
+in_query <- function(name,
+                     value = NULL,
+                     error_call = caller_env()) {
+  check_string(name, call = error_call)
+  out <- list(value = value, name = name)
+  class(out) <- c("httr2_query_param", "httr2_param")
+  out
+}
+
+#' @rdname in_query
+#' @export
+in_header <- function(name,
```

Comment:
I didn't encounter an example where pagination information is passed via the header but this probably exists (and also makes sense to me)

Reply:
https://docs.github.com/en/rest/guides/using-pagination-in-the-rest-api?apiVersion=2022-11-28#using-link-headers maybe?